### PR TITLE
Create Blog “federating-spire-on-hpe-greenlake-for-private-cloud-enterprise”

### PR DESCRIPTION
**Pull request by akanshasajimon**
Automatically generated by Netlify CMS

### CODE REVIEW

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+> '.dynamicActiveSecrets[0].secret.tlsCertificate.certificateChain.inlineBytes' | base64 --decode > chain.pem 
+
+Cluster1:~ # openssl x509 -in chain.pem -text | grep SPIRE 
+
+        Subject: C = US, O = SPIRE, x500UniqueIdentifier = c3dd29b5f4a326f8a567854407456ea9 
+```
+
+*Note: Similarly, certificates for other pods can be checked.* 
+
+## 4﻿.5 Create a secret to hold bookinfo product page credentials like this:
```
Comment (drensing) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:277]: This is an instance similar to the one shown earlier where this appears to be the instructions that you are giving the reader as opposed to an actual header. I'd recommend creating an actual header first that says something like "Create a secret" and then place this, in regular font, as a sentence below it. This way you maintain a consistent use of 1) header 2) instructions.

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+
+Obtain the ca cert using this command: 
+
+```shellsession
+kubectl exec -n spire -c spire-server deployment/spire-server -- /opt/spire/bin/spire-server bundle list
+```
+
+*Note: Copy the ca cert under the cluster1.demo section into a new file bookinfo.ca.cert.* 
+
+Apply the gateway configuration for the bookinfo application found here: (It uses the istio-ingress gateway)
```
Comment (drensing) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:305]: Are you missing a link? You say "found here". That usually indicates a hot-link. Perhaps you meant "for the bookinfo application shown below:" instead?

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+
+*Note: Copy the ca cert under the cluster1.demo section into a new file bookinfo.ca.cert.* 
+
+Apply the gateway configuration for the bookinfo application found here: (It uses the istio-ingress gateway) 
+
+```shellsession
+kubectl apply -f services/istio/release-1.17/bookinto-gateway.yaml
+```
+
+## 4﻿.6 Generate traffic to the sample application on cluster2 from the curl greeter on cluster 1:
```
Comment (drensing) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:311]: Remove the colon at the end of this line

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+
+Obtain ca cert using: 
+
+```shellsession
+kubectl exec -n spire -c spire-server deployment/spire-server -- /opt/spire/bin/spire-server bundle list
+```
+
+*Note: Copy the ca cert under the cluster2.demo section into a new file bookinfo.ca.cert.* 
+
+## 4﻿.7 Visualize using Service Mesh:
```
Comment (drensing) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:433]: Remove the colon at the end of this line

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+
+```shellsession
+kubectl exec -n spire -c spire-server deployment/spire-server -- /opt/spire/bin/spire-server bundle list
+```
+
+*Note: Copy the ca cert under the cluster2.demo section into a new file bookinfo.ca.cert.* 
+
+## 4﻿.7 Visualize using Service Mesh: 
+
+Using the Kiali dashboard, observing the graphs of generated traffic.
```
Comment (drensing) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:435]: "observing" should be "observe".

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+
+## 4﻿.7 Visualize using Service Mesh: 
+
+Using the Kiali dashboard, observing the graphs of generated traffic. 
+
+The graph below shows services communication, and the locks symbolize mTls protocol. 
+
+![mTLS Graph](/img/mtls.png)
+
+The goal of this blog was to guide you through federating SPIRE across two Kubernetes clusters deployed on HPE GreenLake for Private Cloud Enterprise by creating a cluster federated trust domain and federated ClusterSpiffeIDs for your sample application workloads as well as to help you visualize your service mesh through Kiali Dashboard.
```
Comment (drensing) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:441]: I'd recommend placing a header above this paragraph that says "Summary"

Change "blog" to "blog post". I would also break this paragraph out into at least two sentences. I suggest the following:
The goal of this blog post was to guide you through federating SPIRE across two Kubernetes clusters deployed on HPE GreenLake for Private Cloud Enterprise. **It shows you how to do this** by creating a cluster federated trust domain and federated ClusterSpiffeIDs for your sample application workloads and **then helps** you visualize your service mesh through Kiali Dashboard.

**Review by drensing — COMMENTED** [review]
Hi! Just a few things to watch for... try to retain consistency in formatting throughout the post. Headers don't use colons, in general.

**Comment by akanshasajimon** [comment]
Hi @drensing , thanks for all your recommendations. I've addressed all your comments, please let me know if there are any more edits required.

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+
+## 2.3 Patch Istio Ingress Gateway
+
+### 2﻿.3.1 Apply SPIFFE ID
+
+First, you must get and apply one of SPIRE controller manager’s [CRD (Custom Resource Definition)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) ClusterSPIFFEID. The CRD - ClusterSPIFFEID is a cluster-wide resource used to register workloads with SPIRE. The ClusterSPIFFEID can target all workloads in the cluster or can be optionally scoped to specific pods or namespaces via label selectors.  
+
+Create a ClusterSPIFFEID CRD to generate registration entries in SPIRE server for all workloads labeled ***spiffe.io/spire-managed-identity: true.***
+
+Get the ClusterSPIFFEID used for this demo using this [link](https://github.com/cxteamtrials/caas-trials-content/blob/main/services/spire/clusterspiffeid-example.yaml) and  apply it to both clusters.
```
Comment (choukden) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:107]: I am not sure I fully understand how one will get the ClusterSPIFFEID using the yaml file https://github.com/cxteamtrials/caas-trials-content/blob/main/services/spire/clusterspiffeid-example.yaml ?

In the sentence "Get the ClusterSPIFFEID used for this demo using this [link](https://github.com/cxteamtrials/caas-trials-content/blob/main/services/spire/clusterspiffeid-example.yaml) and  apply it to both clusters", it sounds to me there are two actions here: 
1- get the ClusterSPIFFIDID 
2- apply the ClusterSPIFFEID to each demo cluster.

However, there is only one command mentioned in the section 2.3.1 Apply SPIFFE ID:
kubectl apply -f services/spire/clusterspiffeid-example.yaml

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+istiod-d5bc8669c-thbpj                 1/1     Running   0          37d
+```
+
+# Step 3. Federating SPIRE
+
+## 3﻿.1 Expose SPIRE server bundle endpoint
+
+Assign an external IP to your spire-server-bundle-endpoint service on each cluster.  
+
+SPIFFE (*Secure Production Identity Framework For Everyone*) is a specification for implementing identity for workloads, and SPIRE is the code that implements this specification in practice. A SPIFFE bundle is a resource that contains the public key material needed to authenticate credentials from a particular trust domain. A SPIFFE bundle endpoint is a resource (represented by a URL) that serves a copy of a SPIFFE bundle for a trust domain. SPIFFE control planes may both expose and consume these endpoints to transfer bundles between themselves, thereby achieving federation. The SPIRE server is used to host the “spire-server-bundle-endpoint” service that serves the SPIFFE bundle to an external Spire agent of a different trust domain.
```
Comment (choukden) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:147]: Last sentence:
...to an external SPIRE agent of a different trust domain.

For consistency, I suggest spelling SPIRE agent with "SPIRE" in capital letters.

File: content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+
+In Cluster 1, apply a new ClusterSpiffeID called ***federated*** that registers resources with the label **spiffe.io/spire-managed-identity=curl-greeter** that can be federated with cluster2. Create a resource called ***curl-greeter*** that has the label: **spiffe.io/spire-managed-identity=curl-greeter** and annotation: **inject.istio.io/templates=sidecar, spire** 
+
+```shellsession
+kubectl run curl-greeter --image=radial/busyboxplus:curl --labels="spiffe.io/spire-managed-identity=curl-greeter" --overrides='{ "apiVersion": "v1", "spec": { "template": {"metadata": {"annotations": { "inject.istio.io/templates":"sidecar,spire" } } }}}' -i --tty 
+```
+
+## 4﻿.2 Deploy Bookinfo Sample Application in Cluster-2
+
+In Cluster 2, apply a new ClusterSpiffeID called ***federated*** that registers resources with the label **spiffe.io/spire-managed-identity=spire** that can be federated with cluster1. Apply the bookinfo sample application manifest.
```
Comment (choukden) [inline content/blog/federating-spire-on-hpe-greenlake-for-private-cloud-enterprise.md:234]: For Cluster 2, you mention in Cluster 2, apply a new ClusterSpiffeID called federated that registers resouces with the label ....

Does it mean that one would need to apply the same kubectl command as in 4.1 before applying the Bookinfo sample app manifest in Cluster 2?
kubectl run curl-greeter --image=radial/busyboxplus:curl --labels="spiffe.io/spire-managed-identity=curl-greeter" --overrides='{ "apiVersion": "v1", "spec": { "template": {"metadata": {"annotations": { "inject.istio.io/templates":"sidecar,spire" } } }}}' -i --tty

**Comment by akanshasajimon** [comment]
Hi Denis,
Thanks for the review, I've incorporated your suggestions.
With regards,
Akansha

**Pull request merged**